module_utils/certificate/providers/base.py

```diff
@@ -0,0 +1,291 @@
+import os
+
+from abc import ABCMeta, abstractmethod
+
+from cryptography import x509
+from cryptography.hazmat.backends import default_backend
+from cryptography.x509.oid import NameOID
+
+from ansible.module_utils import six
+
+if six.PY2:
+    FileNotFoundError = IOError  # pylint: disable=redefined-builtin
+
+
+class CertificateProxy:
+    """Proxy class that represents certificate-like objects.
+
+    The CertificateProxy can represent both existing
+    certificate and respective CSR and also new certificate
+    requests originated from an Ansible module.
+
+    This class only knows about certificate attributes that
+    can be useful for idempotency purposes of a certificate.
+    """
+
+    def __init__(self):
+        self.cert_data = {}
+        self._x509_obj = None
+
+    @classmethod
+    def load_from_params(cls, params):
+        """Return a CertificateProxy object initialized from a dict."""
+        # pylint: disable=protected-access
+        cert_like = cls()
+
+        map_attrs = ["dns", "ip"]
+        info = {k: v for k, v in params.items() if k in map_attrs}
+
+        info["common_name"] = cert_like._get_common_name_from_params(params)
+
+        cert_like.cert_data = info
+        return cert_like
+
+    @staticmethod
+    def _get_common_name_from_params(params):
+        """Infer the best `common_name` from given parameters.
+
+        If `common_name` is not provided use the first `dns` or first `ip`.
+        """
+        common_name = params.get("common_name")
+        if common_name is None:
+            dns = params.get("dns")
+            if dns:
+                common_name = dns[0]
+
+            ip = params.get("ip")
+            if ip:
+                common_name = ip[0]
+
+        return common_name
+
+    @classmethod
+    def load_from_pem(cls, cert_pem=None, csr_pem=None):
+        """Return a CertificateProxy object initialized from PEM data.
+
+        The argument `cert_pem` it's a certificate in PEM format. The argument
+        `csr_pem` it's a CSR in PEM format.
```

Comment:
use `is` instead of `it's` here

Reply:
4e74b07b67c935b8d577375e7779606444905190